examples/post-sync/controller_test.go

```diff
@@ -0,0 +1,227 @@
+					},
+					Status: v1beta1.RootSyncStatus{
+						Status: v1beta1.Status{
+							Source: v1beta1.SourceStatus{
+								Commit: "abc123",
+								Errors: []v1beta1.ConfigSyncError{
+									{
+										Code:         "1234",
+										ErrorMessage: "test error",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedLogs:   true,
+			expectedError:  false,
+			expectedCommit: "abc123",
+		},
+		{
+			name:     "RepoSync with error",
+			syncKind: "RepoSync",
+			setupObjects: []client.Object{
+				&v1beta1.RepoSync{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "test-reposync",
+						Namespace: "test-namespace",
+					},
+					Status: v1beta1.RepoSyncStatus{
+						Status: v1beta1.Status{
+							Source: v1beta1.SourceStatus{
+								Commit: "def456",
+								Errors: []v1beta1.ConfigSyncError{
+									{
+										Code:         "5678",
+										ErrorMessage: "test error",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedLogs:   true,
+			expectedError:  false,
+			expectedCommit: "def456",
+		},
+		{
+			name:     "RootSync without error",
+			syncKind: "RootSync",
+			setupObjects: []client.Object{
+				&v1beta1.RootSync{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "test-rootsync-success",
+						Namespace: "config-management-system",
+					},
+					Status: v1beta1.RootSyncStatus{
+						Status: v1beta1.Status{
+							Source: v1beta1.SourceStatus{
+								Commit: "ghi789",
+							},
+							Rendering: v1beta1.RenderingStatus{
+								Commit: "ghi789",
+							},
+							Sync: v1beta1.SyncStatus{
+								Commit: "ghi789",
+							},
+						},
+					},
+				},
+			},
+			expectedLogs:   false,
+			expectedError:  false,
+			expectedCommit: "ghi789",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// Create a fake client with the test objects
+			client := fake.NewClientBuilder().
+				WithScheme(s).
+				WithObjects(tt.setupObjects...).
+				Build()
+
+			// Create a test logger
+			logger := testr.New(t)
+
+			// Create a status tracker
+			statusTracker := NewStatusTracker()
+
+			// Create the controller
+			controller := NewSyncStatusController(client, logger, statusTracker, tt.syncKind)
+
+			// Create the request
+			var req reconcile.Request
+			switch tt.syncKind {
+			case "RootSync":
+				req = reconcile.Request{
+					NamespacedName: types.NamespacedName{
+						Name:      "test-rootsync",
+						Namespace: "config-management-system",
+					},
+				}
+				if tt.name == "RootSync without error" {
+					req.Name = "test-rootsync-success"
+				}
+			case "RepoSync":
+				req = reconcile.Request{
+					NamespacedName: types.NamespacedName{
+						Name:      "test-reposync",
+						Namespace: "test-namespace",
+					},
+				}
+			}
+
+			// Run the reconciliation
+			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancel()
+
+			_, err := controller.Reconcile(ctx, req)
```

Comment:
FYI, when you do reconcile tests like this, it only tests the Reconcile method, not the For/Watch registration or any of the other setup code in main. 

If you decide you want to test more of it in integration tests, you can spin up a fake client and a controller-manager and run it in the background while making API calls. That's how we do a integration testing for the reconciler: https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/pkg/reconcilermanager/controllers/rootsync_controller_manager_test.go

Reply:
Let me see how big of a change it is, we can always refine this in a later PR too.

Reply:
mmm setting up the fake client infrastructure looks like it'll take more than a quick fix. Since we already have this behavior covered in the e2e tests, shall we unblock the PR for now and I'll tackle the fake client setup as a follow-up task.

Reply:
yeah, i dont think integration testing this is critical for MVP. Just wanted you to know. If we make this more of a product later it might be worth doing later.